lib/puppet/util/monkey_patches.rb

```diff
@@ -3,6 +3,10 @@
 module Puppet::Util::MonkeyPatches
 end
 
+require 'sorted_set'
```

Comment:
One option is to [not add `sorted_set`](https://github.com/puppetlabs/puppet/pull/8589/commits/3fcb9875273dcc01a33f3fc1b47b1129d00f0733#diff-aea1a1bf6b5975fe75d58571ef069e7a742a88fa2d49c4739e346a2614329050) as a runtime dependency. Then we add a soft dependency check to detect if the gem is missing on Ruby 3.

However, that would mean any gem that depends on puppet would have to know about puppet's dependencies, leading to dependency hell.

If we keep the runtime dependency on `sorted_set`, then the dependency needs to be added to puppet-runtime before this can be merged, like we did for [`scanf`](https://tickets.puppetlabs.com/browse/PA-3350). There will also be downstream affects for puppetserver, bolt, pdk, etc.

Reply:
> However, that would mean any gem that depends on puppet (like rspec-puppet) would have to know about puppet's dependencies, leading to gem dependency hell.

I'm for explicitly declaring our dependencies and keep them in control.

I think adding the gem to puppet-runtime is the best we can do(i don't see a way to add the gem only on Ruby > 2). In the end, it will still end up there as we'll be moving to ruby3

Reply:
So I think I can avoid this mess by using our RbTreeMap class instead of SortedSet. Eventually we'll probably want to move to the native `rbtree` gem, but we can cross that bridge another day.